File: tools/vision-ocr-batch.mjs
// vision-ocr-batch.mjs
// 用 vision-bridge（DashScope Qwen-VL）批量重跑图片版 PDF 的 OCR
// 用法: node tools/vision-ocr-batch.mjs <书名key> [起始页] [结束页]
// 书名key: 1haoshu | 3haoshu | 5haoshu | 6haoshu | qimen10 | qimen11 | kapai
// 输出: tools/extracted-vision/<书名>.ocr.txt （含 ===== 第 N 页 ===== 分隔符）
// 缓存: <cwd>/.ai/vision/*.md （复用 vision-bridge 的缓存格式，同图重复免付费）

import { spawnSync } from 'child_process';
import fs from 'fs';
import path from 'path';
import os from 'os';
import { createHash } from 'crypto';

// ===== 配置 =====
const ROOT = 'C:/1xiangmu/yixiao';
const OUT_DIR = path.join(ROOT, 'tools/extracted-vision');
const TMP_DIR = path.join(os.tmpdir(), 'vocr');
const PPM = 'C:/Users/1/AppData/Local/Microsoft/WinGet/Packages/oschwartz10612.Poppler_Microsoft.Winget.Source_8wekyb3d8bbwe/poppler-25.07.0/Library/bin/pdftoppm.exe';
const CONCURRENCY = 3; // 并发视觉调用数（实测 3 最稳）
const DPI = 150;

const BOOKS = {
  '1haoshu':  { pdf: 'materials/lectures/1号书26齐德龙技术实物防火篇.pdf', pages: 314, name: '1号书26齐德龙技术实物防火篇' },
  '3haoshu':  { pdf: 'materials/lectures/3号书设施篇.pdf', pages: 447, name: '3号书设施篇' },
  '5haoshu':  { pdf: 'materials/lectures/5号书消防综合能力内页.pdf', pages: 347, name: '5号书消防综合能力内页' },
  '6haoshu':  { pdf: 'materials/exercises/6号综合习题.pdf', pages: 256, name: '6号综合习题' },
  'qimen10':  { pdf: 'materials/qimen-dunjia/2026齐德龙消防10-齐门遁甲横向.pdf', pages: 280, name: '2026齐德龙消防10-齐门遁甲横向' },
  'qimen11':  { pdf: 'materials/qimen-dunjia/2026齐德龙消防11-齐门遁甲竖向.pdf', pages: 395, name: '2026齐德龙消防11-齐门遁甲竖向' },
  'kapai':    { pdf: 'materials/qimen-dunjia/26卡牌大师.pdf', pages: 100, name: '26卡牌大师' },
};

// ===== 读取 vision-bridge 配置（含 API key）=====
function loadConfig() {
  const cfg = JSON.parse(fs.readFileSync('C:/Users/1/.claude.json', 'utf8'));
  // 找 vision-bridge 配置（大写或小写路径键）
  const keys = ['C:/1xiangmu/yixiao', 'c:/1xiangmu/yixiao', 'C:/Users/1'];
  for (const k of keys) {
    const mcp = cfg.projects?.[k]?.mcpServers?.['vision-bridge'];
    if (mcp?.env) return mcp.env;
  }
  throw new Error('未找到 vision-bridge 配置');
}

// ===== DashScope 调用（OpenAI 兼容）=====
async function callVision(dataUrl) {
  const env = loadConfig();
  const baseUrl = env.VISION_BRIDGE_BASE_URL || 'https://dashscope.aliyuncs.com/compatible-mode/v1';
  const models = (env.VISION_BRIDGE_MODELS || 'qwen-vl-max').split(',');
  const apiKey = env.VISION_BRIDGE_API_KEY;
  const prompt = 'Extract ALL visible text verbatim. Preserve line breaks and structure. Output ONLY the text content, no commentary.';

  let lastErr;
  for (const model of models) {
    try {
      const resp = await fetch(`${baseUrl}/chat/completions`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${apiKey}` },
        body: JSON.stringify({
          model,
          temperature: 0.2,
          max_tokens: 4096,
          messages: [{ role: 'user', content: [
            { type: 'text', text: prompt },
            { type: 'image_url', image_url: { url: dataUrl } },
          ]}],
        }),
      });
      if (!resp.ok) {
        lastErr = new Error(`HTTP ${resp.status}: ${await resp.text()}`);
        continue;
      }
      const data = await resp.json();
      return { text: data.choices?.[0]?.message?.content?.trim() || '', model };
    } catch (e) {
      lastErr = e;
    }
  }
  throw lastErr;
}

// ===== 缓存 =====
function cacheKey(source, mode, question) {
  return createHash('sha256').update(`${source}|${mode}|${question ?? ''}`).digest('hex').slice(0, 16);
}
function readCache(dir, key) {
  const f = path.join(dir, `${key}.md`);
  if (!fs.existsSync(f)) return null;
  const content = fs.readFileSync(f, 'utf8');
  const m = content.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
  return m ? m[2] : null;
}
function writeCache(dir, key, source, mode, model, body) {
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(path.join(dir, `${key}.md`), `---\nsource: ${source}\nmode: ${mode}\nmodel: ${model}\n---\n${body}`, 'utf8');
}

// ===== 渲染 PDF 页为 PNG =====
function renderPage(pdf, pageNum, outPath) {
  const base = outPath.replace(/\.png$/, '');
  // pdftoppm 输出为 <base>-NNN.png，需清理旧文件后渲染再定位
  for (const old of fs.existsSync(path.dirname(base)) ? fs.readdirSync(path.dirname(base)) : []) {
    if (old.startsWith(path.basename(base))) {
      try { fs.unlinkSync(path.join(path.dirname(base), old)); } catch {}
    }
  }
  const res = spawnSync(PPM, ['-png', '-r', String(DPI), '-f', String(pageNum), '-l', String(pageNum), pdf, base], { encoding: 'utf8' });
  const match = fs.existsSync(path.dirname(base)) ? fs.readdirSync(path.dirname(base)).find(f => f.startsWith(path.basename(base)) && f.endsWith('.png')) : null;
  return match ? path.join(path.dirname(base), match) : null;
}

// ===== 主流程 =====
async function main() {
  const bookKey = process.argv[2];
  const startPage = parseInt(process.argv[3] || '1', 10);
  const endPage = parseInt(process.argv[4] || '9999', 10);
  const book = BOOKS[bookKey];
  if (!book) { console.error('未知书名key:', bookKey, '可用:', Object.keys(BOOKS).join(',')); process.exit(1); }

  const env = loadConfig();
  Object.assign(process.env, env);
  const VISION_CACHE_DIR = path.join(ROOT, '.ai/vision');
  const bookOut = path.join(OUT_DIR, `${book.name}.ocr.txt`);
  fs.mkdirSync(OUT_DIR, { recursive: true });
  fs.mkdirSync(TMP_DIR, { recursive: true });

  const mode = 'ocr';
  const question = ''; // 固定为空，保证同图缓存命中

  // 已完成的页（断点续跑：检查输出文件里已有的分隔符）
  const donePages = new Set();
  if (fs.existsSync(bookOut)) {
    const existing = fs.readFileSync(bookOut, 'utf8');
    const re = /===== 第 (\d+) 页 =====/g;
    let m;
    while ((m = re.exec(existing))) donePages.add(parseInt(m[1], 10));
  }

  const pageRange = [];
  for (let p = startPage; p <= Math.min(endPage, book.pages); p++) {
    if (!donePages.has(p)) pageRange.push(p);
  }
  console.log(`[${bookKey}] 需处理 ${pageRange.length} 页 (共 ${book.pages} 页, 已完成 ${donePages.size} 页)`);
  if (pageRange.length === 0) { console.log('全部完成'); process.exit(0); }

  const t0 = Date.now();
  let ok = 0, fail = 0;

  // 并发处理（每页：渲染→视觉OCR→写结果）
  let idx = 0;
  async function worker() {
    while (idx < pageRange.length) {
      const page = pageRange[idx++];
      const pngBase = path.join(TMP_DIR, `${bookKey}_p${page}.png`);
      try {
        const png = renderPage(path.join(ROOT, book.pdf), page, pngBase);
        if (!png) { console.error(`页${page} 渲染失败`); fail++; continue; }

        const key = cacheKey(png, mode, question);
        let text = readCache(VISION_CACHE_DIR, key);
        let model = 'cache';
        if (text === null) {
          const bs = fs.readFileSync(png).toString('base64');
          const dataUrl = `data:image/png;base64,${bs}`;
          const res = await callVision(dataUrl);
          text = res.text;
          model = res.model;
          writeCache(VISION_CACHE_DIR, key, png, mode, model, text);
        }

        // 追加到输出文件
        const block = `\n===== 第 ${page} 页 =====\n\n${text}\n`;
        fs.appendFileSync(bookOut, block);
        ok++;
        const elapsed = ((Date.now() - t0) / 1000).toFixed(0);
        const eta = ok > 0 ? (elapsed / ok * (pageRange.length - ok) / 60).toFixed(1) : '?';
        console.log(`[${bookKey}] 页${page}/${book.pages} 完成 (${model}, ${text.length}字) 已用${elapsed}s 预计剩${eta}分钟`);
        fs.unlinkSync(png); // 清理临时 PNG
      } catch (e) {
        console.error(`[${bookKey}] 页${page} 失败: ${e.message}`);
        fail++;
        try { if (fs.existsSync(png)) fs.unlinkSync(png); } catch {}
      }
    }
  }

  await Promise.all(Array.from({ length: CONCURRENCY }, () => worker()));

  const total = ((Date.now() - t0) / 1000).toFixed(0);
  console.log(`[${bookKey}] 完成: 成功${ok} 失败${fail} 总耗时${total}s`);
}

main().catch(e => { console.error('致命错误:', e.message); process.exit(1); });

File: tools/vision-ocr-batch.py
#!/usr/bin/env python3
# vision-ocr-batch.py
# 用 PyMuPDF 渲染 PDF 页 + qwen3.7-plus 视觉 OCR 重跑图片版 PDF
# 用法: python tools/vision-ocr-batch.py <书名key> [起始页] [结束页]
# 书名key: 3haoshu | 5haoshu | 6haoshu
# 输出: tools/extracted-vision/<书名>.ocr.txt （含 ===== 第 N 页 ===== 分隔符）
# 缓存: <cwd>/.ai/vision/*.md （与 vision-bridge / 原 node 脚本同格式，同图重复免付费）
# 断点续跑: 检查输出文件已有的页号，跳过已完成页
# 中止保护: 连续 5 次 HTTP 403/429/超时 即中止（额度耗尽），下次续跑

import sys, os, json, hashlib, base64, re, time, threading, urllib.request, urllib.error
from concurrent.futures import ThreadPoolExecutor, as_completed

try:
    import fitz  # PyMuPDF
except ImportError:
    print('请先 pip install PyMuPDF'); sys.exit(1)

ROOT = 'C:/1xiangmu/yixiao'
OUT_DIR = os.path.join(ROOT, 'tools/extracted-vision')
VISION_CACHE_DIR = os.path.join(ROOT, '.ai/vision')
CONCURRENCY = 3
DPI = 150
ABORT_FAIL_STREAK = 5  # 连续失败阈值

BOOKS = {
    '3haoshu': {'pdf': 'materials/lectures/3号书设施篇.pdf', 'pages': 447, 'name': '3号书设施篇'},
    '5haoshu': {'pdf': 'materials/lectures/5号书消防综合能力内页.pdf', 'pages': 347, 'name': '5号书消防综合能力内页'},
    '6haoshu': {'pdf': 'materials/exercises/6号综合习题.pdf', 'pages': 256, 'name': '6号综合习题'},
}

def load_config():
    cfg = json.load(open('C:/Users/30828/.claude.json', 'r', encoding='utf-8'))
    for p in cfg.get('projects', {}):
        m = cfg['projects'][p].get('mcpServers', {}).get('vision-bridge')
        if m and m.get('env'): return m['env']
    raise Exception('未找到 vision-bridge 配置')

ENV = load_config()
BASE_URL = ENV.get('VISION_BRIDGE_BASE_URL', 'https://dashscope.aliyuncs.com/compatible-mode/v1')
MODELS = [m for m in (ENV.get('VISION_BRIDGE_MODELS', 'qwen-vl-max')).split(',') if not re.search(r'realtime', m, re.I)]
API_KEY = ENV['VISION_BRIDGE_API_KEY']

PROMPT = 'Extract ALL visible text verbatim. Preserve line breaks and structure. Output ONLY the text content, no commentary.'

# 中止信号
abort_event = threading.Event()
fail_streak = 0
streak_lock = threading.Lock()
write_lock = threading.Lock()

def cache_key(source, mode, question):
    return hashlib.sha256(f'{source}|{mode}|{question or ""}'.encode()).hexdigest()[:16]

def read_cache(key):
    f = os.path.join(VISION_CACHE_DIR, f'{key}.md')
    if not os.path.exists(f): return None
    content = open(f, 'r', encoding='utf-8').read()
    m = re.match(r'^---\n[\s\S]*?\n---\n([\s\S]*)$', content)
    return m.group(1) if m else None

def write_cache(key, source, mode, model, body):
    os.makedirs(VISION_CACHE_DIR, exist_ok=True)
    f = os.path.join(VISION_CACHE_DIR, f'{key}.md')
    open(f, 'w', encoding='utf-8').write(f'---\nsource: {source}\nmode: {mode}\nmodel: {model}\n---\n{body}')

def render_page(pdf_path, page_num):
    doc = fitz.open(pdf_path)
    page = doc[page_num - 1]
    mat = fitz.Matrix(DPI / 72, DPI / 72)
    pix = page.get_pixmap(matrix=mat)
    png = pix.tobytes('png')
    doc.close()
    return png

def call_vision(png_bytes):
    data_url = 'data:image/png;base64,' + base64.b64encode(png_bytes).decode()
    body = json.dumps({
        'model': None, 'temperature': 0.2, 'max_tokens': 4096,
        'messages': [{'role': 'user', 'content': [
            {'type': 'text', 'text': PROMPT},
            {'type': 'image_url', 'image_url': {'url': data_url}},
        ]}],
    })
    last_err = None
    for model in MODELS:
        payload = body.replace('"model": null', f'"model": "{model}"')
        try:
            req = urllib.request.Request(BASE_URL + '/chat/completions', data=payload.encode('utf-8'),
                                         headers={'Content-Type': 'application/json', 'Authorization': f'Bearer {API_KEY}'})
            with urllib.request.urlopen(req, timeout=180) as resp:
                data = json.loads(resp.read().decode('utf-8'))
                text = (data.get('choices', [{}])[0].get('message', {}).get('content') or '').strip()
                if text: return {'text': text, 'model': data.get('model', model)}
                last_err = f'模型 {model} 返回空内容'
        except urllib.error.HTTPError as e:
            last_err = f'HTTP {e.code}: {e.read().decode("utf-8", "ignore")[:300]}'
            # 403/429 视为额度/限流，记入连续失败
            if e.code in (403, 429): raise
        except Exception as e:
            last_err = f'{type(e).__name__}: {e}'
            raise  # 超时等也记入连续失败
    raise Exception(last_err or '所有模型均失败')

def process_page(book_key, book, page, book_out):
    global fail_streak
    if abort_event.is_set(): return ('skip', page)
    pdf_path = os.path.join(ROOT, book['pdf'])
    source = f'{book["name"]}_p{page}'
    key = cache_key(source, 'ocr', '')
    try:
        text = read_cache(key)
        model = 'cache'
        if text is None:
            png = render_page(pdf_path, page)
            res = call_vision(png)
            text = res['text']; model = res['model']
            write_cache(key, source, 'ocr', model, text)
        block = f'\n===== 第 {page} 页 =====\n\n{text}\n'
        with write_lock:
            with open(book_out, 'a', encoding='utf-8') as f: f.write(block)
        with streak_lock:
            fail_streak = 0
        return ('ok', page, model, len(text))
    except Exception as e:
        with streak_lock:
            fail_streak += 1
            cur = fail_streak
        if cur >= ABORT_FAIL_STREAK:
            abort_event.set()
        return ('fail', page, str(e)[:200], cur)

def main():
    book_key = sys.argv[1] if len(sys.argv) > 1 else ''
    start_page = int(sys.argv[2]) if len(sys.argv) > 2 else 1
    end_page = int(sys.argv[3]) if len(sys.argv) > 3 else 9999
    book = BOOKS.get(book_key)
    if not book: print(f'未知书名key: {book_key} 可用: {",".join(BOOKS)}'); sys.exit(1)

    os.makedirs(OUT_DIR, exist_ok=True)
    book_out = os.path.join(OUT_DIR, f'{book["name"]}.ocr.txt')

    # 断点续跑
    done = set()
    if os.path.exists(book_out):
        for m in re.finditer(r'===== 第 (\d+) 页 =====', open(book_out, 'r', encoding='utf-8').read()):
            done.add(int(m.group(1)))

    pages = [p for p in range(start_page, min(end_page, book['pages']) + 1) if p not in done]
    print(f'[{book_key}] 需处理 {len(pages)} 页 (共 {book["pages"]} 页, 已完成 {len(done)} 页)', flush=True)
    if not pages: print('全部完成'); return

    t0 = time.time(); ok = 0; fail = 0
    with ThreadPoolExecutor(max_workers=CONCURRENCY) as ex:
        futs = {ex.submit(process_page, book_key, book, p, book_out): p for p in pages}
        for fut in as_completed(futs):
            r = fut.result()
            if r[0] == 'ok':
                ok += 1
                elapsed = time.time() - t0
                eta = (elapsed / ok * (len(pages) - ok) / 60) if ok else 0
                print(f'[{book_key}] 页{r[1]}/{book["pages"]} 完成 ({r[2]}, {r[3]}字) 已用{elapsed:.0f}s 预计剩{eta:.1f}分钟', flush=True)
            elif r[0] == 'fail':
                fail += 1
                print(f'[{book_key}] 页{r[1]} 失败(连续{r[3]}): {r[2]}', flush=True)
                if abort_event.is_set():
                    print(f'[{book_key}] 连续失败达{ABORT_FAIL_STREAK}次，中止。已成功{ok}失败{fail}，可断点续跑', flush=True)
                    for f in futs:
                        if not f.done(): f.cancel()
                    break
    print(f'[{book_key}] 结束: 成功{ok} 失败{fail} 总耗时{time.time()-t0:.0f}s', flush=True)

if __name__ == '__main__':
    main()
